File: supabase/functions/_shared/ai-providers.ts
/** Provedores de IA compartilhados — mesma cadeia de fallback dos projetos NaIntegra. */
export const corsHeaders = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers":
    "authorization, x-client-info, apikey, content-type, x-supabase-client-platform, x-supabase-client-platform-version, x-supabase-client-runtime, x-supabase-client-runtime-version",
};

export type AiProvider = "groq" | "github" | "gemini" | "claude" | "openrouter";

export const AI_PRIORITY: readonly AiProvider[] = ["groq", "gemini", "github", "openrouter", "claude"];

export const AI_MODELS: Record<AiProvider, string> = {
  groq: Deno.env.get("GROQ_MODEL") || "llama-3.3-70b-versatile",
  github: Deno.env.get("GITHUB_MODEL") || "openai/gpt-4o-mini",
  gemini: Deno.env.get("GEMINI_MODEL") || "gemini-2.5-flash",
  claude: Deno.env.get("CLAUDE_MODEL") || "claude-sonnet-4-20250514",
  openrouter: Deno.env.get("OPENROUTER_MODEL") || "meta-llama/llama-3.3-70b-instruct",
};

export interface ChatMessage {
  role: "user" | "assistant" | "system";
  content: string;
}

export function parseAiProvider(value: unknown): AiProvider {
  if (value === "groq" || value === "github" || value === "gemini" || value === "claude" || value === "openrouter") {
    return value;
  }
  return "groq";
}

export function listConfiguredProviders(): { provider: AiProvider; model: string; configured: boolean }[] {
  const keys: Record<AiProvider, string> = {
    groq: "GROQ_API_KEY",
    github: "GITHUB_TOKEN",
    gemini: "GEMINI_API_KEY",
    claude: "ANTHROPIC_API_KEY",
    openrouter: "OPENROUTER_API_KEY",
  };
  return AI_PRIORITY.map((provider) => ({
    provider,
    model: AI_MODELS[provider],
    configured: Boolean(Deno.env.get(keys[provider])),
  }));
}

/** Modelos preferidos no Groq, em ordem — a lista da Groq muda com frequência. */
const GROQ_PREFERRED = [
  "llama-3.3-70b-versatile",
  "openai/gpt-oss-120b",
  "moonshotai/kimi-k2-instruct",
  "qwen/qwen3-32b",
  "llama-3.1-8b-instant",
];

let groqModel: string | null = null;

/** Descobre um modelo de chat disponível para a chave (modelos são descontinuados). */
async function discoverGroqModel(key: string): Promise<string> {
  const resp = await fetch("https://api.groq.com/openai/v1/models", {
    headers: { Authorization: `Bearer ${key}` },
  });
  if (!resp.ok) throw new Error(`Groq models ${resp.status}: ${(await resp.text()).slice(0, 120)}`);
  const data = await resp.json();
  const ids: string[] = (data.data || []).map((m: { id: string }) => m.id);
  const pick = GROQ_PREFERRED.find((m) => ids.includes(m)) ||
    ids.find((id) => !/whisper|guard|tts|embed/i.test(id));
  if (!pick) throw new Error("Groq sem modelo de chat disponível");
  return pick;
}

async function groqRequest(model: string, messages: ChatMessage[], maxTokens: number): Promise<Response> {
  const key = Deno.env.get("GROQ_API_KEY")!;
  return fetch("https://api.groq.com/openai/v1/chat/completions", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({ model, max_tokens: maxTokens, messages, temperature: 0.7 }),
  });
}

async function callGroq(messages: ChatMessage[], maxTokens: number): Promise<string> {
  const key = Deno.env.get("GROQ_API_KEY");
  if (!key) throw new Error("GROQ_API_KEY not set");
  let model = groqModel || AI_MODELS.groq;
  let resp = await groqRequest(model, messages, maxTokens);
  if (!resp.ok) {
    const erro = (await resp.text()).slice(0, 200);
    const modeloInvalido = /model_not_found|does not exist|decommissioned|deprecated/i.test(erro);
    if (!modeloInvalido) throw new Error(`Groq ${resp.status}: ${erro}`);
    model = await discoverGroqModel(key);
    resp = await groqRequest(model, messages, maxTokens);
    if (!resp.ok) throw new Error(`Groq ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  }
  groqModel = model;
  const data = await resp.json();
  return data.choices?.[0]?.message?.content || "";
}

async function callGitHub(messages: ChatMessage[], maxTokens: number): Promise<string> {
  const key = Deno.env.get("GITHUB_TOKEN");
  if (!key) throw new Error("GITHUB_TOKEN not set");
  const resp = await fetch("https://models.github.ai/inference/chat/completions", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({
      model: AI_MODELS.github,
      max_tokens: maxTokens,
      messages,
      temperature: 0.7,
    }),
  });
  if (!resp.ok) throw new Error(`GitHub ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const data = await resp.json();
  return data.choices?.[0]?.message?.content || "";
}

async function callGemini(messages: ChatMessage[], maxTokens: number): Promise<string> {
  const key = Deno.env.get("GEMINI_API_KEY");
  if (!key) throw new Error("GEMINI_API_KEY not set");
  const geminiContents = [];
  for (const msg of messages) {
    if (msg.role === "system") {
      geminiContents.push({ role: "user", parts: [{ text: msg.content }] });
      geminiContents.push({ role: "model", parts: [{ text: "Entendido." }] });
    } else {
      geminiContents.push({
        role: msg.role === "assistant" ? "model" : "user",
        parts: [{ text: msg.content }],
      });
    }
  }
  const resp = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/${AI_MODELS.gemini}:generateContent?key=${key}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        contents: geminiContents,
        generationConfig: { temperature: 0.7, maxOutputTokens: maxTokens },
      }),
    },
  );
  if (!resp.ok) throw new Error(`Gemini ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const data = await resp.json();
  return data.candidates?.[0]?.content?.parts?.[0]?.text || "";
}

async function callClaude(messages: ChatMessage[], maxTokens: number): Promise<string> {
  const key = Deno.env.get("ANTHROPIC_API_KEY");
  if (!key) throw new Error("ANTHROPIC_API_KEY not set");
  const systemMsg = messages.find((m) => m.role === "system")?.content || "";
  const chatMsgs = messages
    .filter((m) => m.role !== "system")
    .map((m) => ({ role: m.role as "user" | "assistant", content: m.content }));
  const resp = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": key,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: AI_MODELS.claude,
      max_tokens: maxTokens,
      system: systemMsg,
      messages: chatMsgs,
      temperature: 0.7,
    }),
  });
  if (!resp.ok) throw new Error(`Claude ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const data = await resp.json();
  return data.content?.[0]?.text || "";
}

async function callOpenRouter(messages: ChatMessage[], maxTokens: number): Promise<string> {
  const key = Deno.env.get("OPENROUTER_API_KEY");
  if (!key) throw new Error("OPENROUTER_API_KEY not set");
  const resp = await fetch("https://openrouter.ai/api/v1/chat/completions", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({
      model: AI_MODELS.openrouter,
      max_tokens: maxTokens,
      messages,
      temperature: 0.7,
    }),
  });
  if (!resp.ok) throw new Error(`OpenRouter ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const data = await resp.json();
  return data.choices?.[0]?.message?.content || "";
}

const providerFns: Record<AiProvider, (msgs: ChatMessage[], maxTokens: number) => Promise<string>> = {
  groq: callGroq,
  github: callGitHub,
  gemini: callGemini,
  claude: callClaude,
  openrouter: callOpenRouter,
};

export async function callWithFallback(
  messages: ChatMessage[],
  maxTokens: number,
  preferred?: AiProvider,
): Promise<{ content: string; provider: AiProvider }> {
  const order = preferred
    ? [preferred, ...AI_PRIORITY.filter((p) => p !== preferred)]
    : [...AI_PRIORITY];
  const errors: string[] = [];
  for (const provider of order) {
    try {
      const content = await providerFns[provider](messages, maxTokens);
      if (content) return { content, provider };
    } catch (e) {
      errors.push(`${provider}: ${e instanceof Error ? e.message : String(e)}`);
    }
  }
  throw new Error(`All AI providers failed: ${errors.join(" | ")}`);
}

/** Geração one-shot (prompt único) — usado por scrapers e dashboard. */
export async function aiGenerate(
  prompt: string,
  maxTokens = 4096,
  preferred?: AiProvider,
): Promise<{ content: string; provider: AiProvider }> {
  return callWithFallback(
    [
      { role: "system", content: "Responda em português do Brasil. Seja preciso e objetivo." },
      { role: "user", content: prompt },
    ],
    maxTokens,
    preferred,
  );
}

File: data/prio3_analysis/ai_providers.py
"""Chamada direta aos provedores de IA, sem depender de Edge Function.

Ordem de tentativa: groq → github (GitHub Models) → gemini → openrouter → claude.
As chaves vêm do ambiente ou do .env da raiz do projeto. No GitHub Actions o
provedor `github` funciona com o GITHUB_TOKEN automático, desde que o job
declare `permissions: models: read` — ou seja, não exige chave própria.
"""
from __future__ import annotations

import json
import os
import ssl
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent
REPO = ROOT.parents[1]

PRIORIDADE = ("groq", "gemini", "openrouter", "claude", "ollama", "github")

CHAVES = {
    "groq": "GROQ_API_KEY",
    "gemini": "GEMINI_API_KEY",
    "openrouter": "OPENROUTER_API_KEY",
    "claude": "ANTHROPIC_API_KEY",
    "ollama": "",  # modelo local: disponibilidade é checada no servidor, não por chave
    "github": "GITHUB_TOKEN",
}

OLLAMA_URL = os.environ.get("OLLAMA_URL", "http://localhost:11434")

MODELOS = {
    "groq": "llama-3.3-70b-versatile",
    "ollama": "llama3.2:3b",
    "github": "openai/gpt-4o-mini",
    "gemini": "gemini-2.5-flash",
    "openrouter": "meta-llama/llama-3.3-70b-instruct",
    "claude": "claude-sonnet-4-20250514",
}

# modelos preferidos no Groq — a lista deles muda com frequência
GROQ_PREFERIDOS = (
    "llama-3.3-70b-versatile",
    "openai/gpt-oss-120b",
    "moonshotai/kimi-k2-instruct",
    "qwen/qwen3-32b",
    "llama-3.1-8b-instant",
)

try:
    import certifi

    _SSL = ssl.create_default_context(cafile=certifi.where())
except Exception:  # noqa: BLE001
    _SSL = ssl._create_unverified_context()

_groq_modelo: str | None = None
_env_arquivo: dict[str, str] | None = None


def _do_env(nome: str) -> str | None:
    """Chave do ambiente; se ausente, procura no .env da raiz do repositório."""
    global _env_arquivo
    if os.environ.get(nome):
        return os.environ[nome]
    if _env_arquivo is None:
        _env_arquivo = {}
        for arq in (REPO / ".env", ROOT / ".env"):
            if not arq.is_file():
                continue
            for linha in arq.read_text(encoding="utf-8", errors="replace").splitlines():
                linha = linha.strip()
                if not linha or linha.startswith("#") or "=" not in linha:
                    continue
                k, v = linha.split("=", 1)
                v = v.strip().strip('"').strip("'")
                if v:
                    _env_arquivo.setdefault(k.strip(), v)
    return _env_arquivo.get(nome)


def _post(url: str, payload: dict, headers: dict, timeout: int = 180) -> dict:
    req = urllib.request.Request(
        url,
        data=json.dumps(payload).encode("utf-8"),
        method="POST",
        headers={"Content-Type": "application/json", **headers},
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout, context=_SSL) as r:
            status, corpo = r.status, r.read().decode("utf-8", "replace")
    except urllib.error.HTTPError as e:
        detalhe = e.read().decode("utf-8", "replace")[:200]
        raise RuntimeError(f"{e.code}: {detalhe}") from None
    if not corpo.strip():
        raise RuntimeError(f"{status}: resposta vazia")
    try:
        return json.loads(corpo)
    except json.JSONDecodeError:
        raise RuntimeError(f"{status}: resposta não-JSON: {corpo[:200]}") from None


def _openai_like(
    url: str, key: str, modelo: str, msgs: list[dict], max_tokens: int, extra: dict | None = None
) -> str:
    data = _post(
        url,
        {"model": modelo, "max_tokens": max_tokens, "messages": msgs, "temperature": 0.7},
        {"Authorization": f"Bearer {key}", **(extra or {})},
    )
    return (data.get("choices") or [{}])[0].get("message", {}).get("content") or ""


def _descobre_modelo_groq(key: str) -> str:
    req = urllib.request.Request(
        "https://api.groq.com/openai/v1/models", headers={"Authorization": f"Bearer {key}"}
    )
    with urllib.request.urlopen(req, timeout=30, context=_SSL) as r:
        ids = [m["id"] for m in (json.loads(r.read().decode()).get("data") or [])]
    for m in GROQ_PREFERIDOS:
        if m in ids:
            return m
    livres = [i for i in ids if not any(x in i for x in ("whisper", "guard", "tts", "embed"))]
    if not livres:
        raise RuntimeError("Groq sem modelo de chat disponível")
    return livres[0]


def _groq(msgs: list[dict], max_tokens: int) -> str:
    global _groq_modelo
    key = _do_env("GROQ_API_KEY")
    if not key:
        raise RuntimeError("GROQ_API_KEY ausente")
    url = "https://api.groq.com/openai/v1/chat/completions"
    modelo = _groq_modelo or os.environ.get("GROQ_MODEL") or MODELOS["groq"]
    try:
        out = _openai_like(url, key, modelo, msgs, max_tokens)
    except RuntimeError as e:
        if not any(t in str(e).lower() for t in ("model_not_found", "does not exist", "decommission", "deprecat")):
            raise
        modelo = _descobre_modelo_groq(key)
        out = _openai_like(url, key, modelo, msgs, max_tokens)
    _groq_modelo = modelo
    return out


def _github(msgs: list[dict], max_tokens: int) -> str:
    key = _do_env("GITHUB_TOKEN")
    if not key:
        raise RuntimeError("GITHUB_TOKEN ausente")
    modelo = os.environ.get("GITHUB_MODEL") or MODELOS["github"]
    # o gateway do GitHub responde vazio sem os cabeçalhos REST dele
    cabecalhos = {"Accept": "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28"}
    return _openai_like(
        "https://models.github.ai/inference/chat/completions", key, modelo, msgs, max_tokens, cabecalhos
    )


def _openrouter(msgs: list[dict], max_tokens: int) -> str:
    key = _do_env("OPENROUTER_API_KEY")
    if not key:
        raise RuntimeError("OPENROUTER_API_KEY ausente")
    modelo = os.environ.get("OPENROUTER_MODEL") or MODELOS["openrouter"]
    return _openai_like("https://openrouter.ai/api/v1/chat/completions", key, modelo, msgs, max_tokens)


def _gemini(msgs: list[dict], max_tokens: int) -> str:
    key = _do_env("GEMINI_API_KEY")
    if not key:
        raise RuntimeError("GEMINI_API_KEY ausente")
    modelo = os.environ.get("GEMINI_MODEL") or MODELOS["gemini"]
    conteudos: list[dict] = []
    for m in msgs:
        if m["role"] == "system":
            conteudos.append({"role": "user", "parts": [{"text": m["content"]}]})
            conteudos.append({"role": "model", "parts": [{"text": "Entendido."}]})
        else:
            papel = "model" if m["role"] == "assistant" else "user"
            conteudos.append({"role": papel, "parts": [{"text": m["content"]}]})
    data = _post(
        f"https://generativelanguage.googleapis.com/v1beta/models/{modelo}:generateContent?key={key}",
        {"contents": conteudos, "generationConfig": {"temperature": 0.7, "maxOutputTokens": max_tokens}},
        {},
    )
    partes = ((data.get("candidates") or [{}])[0].get("content") or {}).get("parts") or [{}]
    return partes[0].get("text") or ""


def _ollama_modelos() -> list[str]:
    """Modelos de chat servidos pelo Ollama local (vazio se não estiver rodando)."""
    try:
        with urllib.request.urlopen(f"{OLLAMA_URL}/api/tags", timeout=3) as r:
            nomes = [m.get("name", "") for m in (json.loads(r.read().decode()).get("models") or [])]
    except Exception:  # noqa: BLE001
        return []
    return [n for n in nomes if n and "embed" not in n]


def _ollama(msgs: list[dict], max_tokens: int) -> str:
    disponiveis = _ollama_modelos()
    if not disponiveis:
        raise RuntimeError("Ollama não está rodando")
    preferido = os.environ.get("OLLAMA_MODEL") or MODELOS["ollama"]
    modelo = preferido if preferido in disponiveis else disponiveis[0]
    pede_json = any("APENAS JSON" in m["content"] or "JSON válido" in m["content"] for m in msgs)
    payload = {
        "model": modelo,
        "messages": msgs,
        "stream": False,
        "options": {"temperature": 0.3, "num_predict": max_tokens},
    }
    if pede_json:
        payload["format"] = "json"
    data = _post(f"{OLLAMA_URL}/api/chat", payload, {}, timeout=900)
    return (data.get("message") or {}).get("content") or ""


def _claude(msgs: list[dict], max_tokens: int) -> str:
    key = _do_env("ANTHROPIC_API_KEY")
    if not key:
        raise RuntimeError("ANTHROPIC_API_KEY ausente")
    sistema = next((m["content"] for m in msgs if m["role"] == "system"), "")
    data = _post(
        "https://api.anthropic.com/v1/messages",
        {
            "model": os.environ.get("CLAUDE_MODEL") or MODELOS["claude"],
            "max_tokens": max_tokens,
            "system": sistema,
            "messages": [m for m in msgs if m["role"] != "system"],
            "temperature": 0.7,
        },
        {"x-api-key": key, "anthropic-version": "2023-06-01"},
    )
    return (data.get("content") or [{}])[0].get("text") or ""


_FUNCOES = {
    "groq": _groq,
    "gemini": _gemini,
    "openrouter": _openrouter,
    "claude": _claude,
    "ollama": _ollama,
    "github": _github,
}


def _disponivel(p: str) -> bool:
    return bool(_ollama_modelos()) if p == "ollama" else bool(_do_env(CHAVES[p]))


def configurados() -> list[str]:
    return [p for p in PRIORIDADE if _disponivel(p)]


def somente_local() -> bool:
    """True quando o único provedor é o modelo local (limita o tamanho dos lotes)."""
    return configurados() == ["ollama"]


def chat(msgs: list[dict], max_tokens: int = 2048, preferido: str | None = None) -> tuple[str, str]:
    """Tenta os provedores em ordem; devolve (conteúdo, provedor)."""
    ordem = ([preferido] if preferido in _FUNCOES else []) + [p for p in PRIORIDADE if p != preferido]
    erros = []
    for p in ordem:
        if not _disponivel(p):
            continue
        try:
            out = _FUNCOES[p](msgs, max_tokens)
            if out.strip():
                return out, p
            erros.append(f"{p}: resposta vazia")
        except Exception as e:  # noqa: BLE001
            erros.append(f"{p}: {e}")
    raise RuntimeError("nenhum provedor respondeu — " + " | ".join(erros or ["sem chaves configuradas"]))


def gerar(prompt: str, max_tokens: int = 2048, preferido: str | None = None) -> tuple[str, str]:
    return chat(
        [
            {"role": "system", "content": "Responda em português do Brasil. Seja preciso e objetivo."},
            {"role": "user", "content": prompt},
        ],
        max_tokens,
        preferido,
    )
